warehouse/malware/checks/setup_patterns/check.py

```diff
@@ -0,0 +1,120 @@
+        if zipfile.is_zipfile(archive_stream):
+            with zipfile.ZipFile(archive_stream) as zipobj:
+                for name in zipobj.namelist():
+                    parts = os.path.split(name)
+                    if len(parts) == 2 and parts[1] == "setup.py":
+                        return zipobj.read(name)
+                return None
+        else:
+            # NOTE: is_zipfile doesn't rewind the fileobj it's given.
+            archive_stream.seek(0)
+
+            # NOTE: We don't need to perform a sanity check on
+            # the (presumed) tarfile's compression here, since we're
+            # extracting from a stream that's already gone through
+            # upload validation.
+            # See _is_valid_dist_file in forklift/legacy.py.
+            try:
+                with tarfile.open(fileobj=archive_stream) as tarobj:
+                    member = tarobj.next()
+                    while member:
+                        parts = os.path.split(member.name)
+                        if len(parts) == 2 and parts[1] == "setup.py":
+                            return tarobj.extractfile(member).read()
+                        member = tarobj.next()
+                    return None
+            except tarfile.TarError:
+                return None
+
+    def scan(self, **kwargs):
+        file_id = kwargs.get("obj_id")
+        file_url = kwargs.get("file_url")
+        if file_id is None or file_url is None:
+            # TODO: Maybe raise here, since the absence of these
+            # arguments is a use/user error.
+            return
+
+        file = self.db.query(File).get(file_id)
+        if file.packagetype != "sdist":
+            # Per PEP 491: bdists do not contain setup.py.
+            # This check only scans dists that contain setup.py, so
+            # we have nothing to perform.
+            return
+
+        # TODO(ww): Handle connection errors, timeouts?
+        # The check harness will catch everything eventually,
+        # so these might not be worth handling explicitly.
+        file_request = requests.get(file_url)
+        if not file_request.ok:
+            # TODO: Indeterminate verdict here?
+            return
+
+        setup_py_contents = self._extract_setup_contents(BytesIO(file_request.content))
+        if setup_py_contents is None:
+            self.add_verdict(
+                file_id=file_id,
+                classification=VerdictClassification.Indeterminate,
+                confidence=VerdictConfidence.High,
+                message="sdist does not contain a suitable setup.py for analysis",
+            )
+            return
+
+        matches = self._yara_rules.match(data=setup_py_contents)
+        for match in matches:
+            self.add_verdict(
+                file_id=file_id,
+                classification=VerdictClassification(match.meta["classification"]),
+                confidence=VerdictConfidence(match.meta["confidence"]),
+                message=match.rule,
```

Comment:
For review: I could make the message here lengthier (e.g. by adding more information to the `meta` section of each YARA rule). Thoughts?

Reply:
looks like the message would currently be something like `process_spawn_in_setup` or `networking_in_setup `? If so that's _probably_ sufficient but something human readable would be nice.

Reply:
> looks like the message would currently be something like `process_spawn_in_setup` or `networking_in_setup `?

Yep, exactly. I'll go ahead and add brief descriptions via `meta` as well. It also probably wouldn't be difficult to convert the match's internal `offset, pattern` list into a `dict` for `MalwareVerdict.details`, if desired.

Reply:
I think leveraging `details` here is a great way to showcase how it is useful :)